stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+
+    // Extra Tag (if any) holds upper bits of case value
+    auto extraTagSize = getSize() - payloadSize;
+    unsigned extraTag = 0;
+    if (extraTagSize > 0) {
+      auto extraTagAddress = address + payloadSize;
+      if (!reader.readInteger(extraTagAddress, extraTagSize,
+                              &extraTag)) {
+        return false;
+      }
+    }
+
+    // The `payloadTagMask` is a subset of the spare bits
+    // where we encode the rest of the case value.
+    auto payloadTagMask = getMultiPayloadTagBitsMask();
+    auto numPayloadTagBits = payloadTagMask.countSetBits();
+    uint64_t payloadTag = 0;
+    if (!payloadTagMask.readMaskedInteger(reader, address, &payloadTag)) {
+      return false;
+    }
+
+    // Combine the extra tag and payload tag info:
+    int tagValue = 0;
+    if (numPayloadTagBits >= 32) {
+      tagValue = payloadTag;
+    } else {
+      tagValue = (extraTag << numPayloadTagBits) | payloadTag;
+    }
+
+    // If the above identifies a payload case, we're done
+    if (static_cast<unsigned>(tagValue) < NumPayloadCases) {
+      *CaseIndex = tagValue;
+      return true;
+    }
+
+    // Otherwise, combine with other payload data to select a non-payload case
+    auto occupiedBits = spareBitsMask; // Copy
+    occupiedBits.complement();
+
+    auto occupiedBitCount = occupiedBits.countSetBits();
+    uint64_t payloadValue = 0;
+    if (!occupiedBits.readMaskedInteger(reader, address, &payloadValue)) {
+      return false;
+    }
+
+    int ComputedCase = 0;
+    if (occupiedBitCount >= 32) {
+      ComputedCase = payloadValue + NumPayloadCases;
+    } else {
+      ComputedCase = (((tagValue - NumPayloadCases) << occupiedBitCount) |  payloadValue) + NumPayloadCases;
+    }
+
+    if (static_cast<unsigned>(ComputedCase) < getNumCases()) {
+      *CaseIndex = ComputedCase;
       return true;
     } else {
-      // Multipayload enums that don't use a separate tag never
-      // export XIs.
+      *CaseIndex = -1;
       return false;
     }
   }
+
+  // The case value is stored in three pieces:
+  // * A separate "discriminator" tag appended to the payload (if necessary)
+  // * A "payload tag" that uses (a subset of) the spare bits
+  // * The remainder of the payload bits (for non-payload cases)
+  // This computes the bits used for the payload tag.
+  BitMask getMultiPayloadTagBitsMask() const {
+    auto payloadTagValues = getNumPayloadCases() - 1;
+    if (getNumCases() > getNumPayloadCases()) {
+      payloadTagValues += 1;
+    }
+    int payloadTagBits = 0;
+    while (payloadTagValues > 0) {
+      payloadTagValues >>= 1;
+      payloadTagBits += 1;
+    }
+    BitMask payloadTagBitsMask = spareBitsMask;
+    payloadTagBitsMask.keepOnlyMostSignificantBits(payloadTagBits);
+    return payloadTagBitsMask;
   }
-  return false;
+};
+
+// XXX FIXME XXX Make sure the following calculation returns `false` (failure)
+// when it is unable to accurately obtain the spare bit mask.
```

Comment:
I'm not yet sure whether it's always possible for the runtime to do this calculation.  However, as long as this code is honest about it's capabilities and returns `false` when needed information is unavailable, everything else here should work.  The question is not really whether we can correctly support all MPEs, but what subset of MPEs we can correctly support.

Reply:
Backed out for this PR.  A future PR will implement this by using additional data from the compiler.